luna/gateware/platform/numato_opsis.py

```diff
@@ -0,0 +1,208 @@
+
+        # Button
+        *ButtonResources(pins="Y3", attrs=Attrs(IOSTANDARD="LVCMOS15", PULLUP="TRUE")),
+
+        # ULPI phy with input clock operation (ULPI clk generated by link)
+        ULPIResource("ulpi",
+            data_sites="Y14 W14 Y18 W17 AA14 AB14 Y16 W15", clk_site="-",
+            dir_site="W13", nxt_site="V13", stp_site="W18", reset_site="V17",
+        ),
+
+        Resource("forward_ulpi_clk", 0, Pins("Y13", dir="o"), Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")),
+        Resource("debug_header", 0, Pins("P28_0:SD_DAT1", dir="o"), Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")),
+        Resource("debug_header", 1, Pins("P28_0:SD_DAT0", dir="o"), Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")),
+        Resource("debug_header", 2, Pins("P28_0:SD_CLK", dir="o"), Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")),
+        Resource("debug_header", 3, Pins("P28_0:SD_CMD", dir="o"), Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")),
+        Resource("debug_header", 4, Pins("P28_0:SD_DAT3", dir="o"), Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")),
+        Resource("debug_header", 5, Pins("P28_0:SD_DAT2", dir="o"), Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")),
+
+    ]
+
+    connectors = [
+        Connector("P28", 0, {
+            "SD_DAT1" : "AB4",
+            "SD_DAT0" : "AA4",
+            "SD_CLK"  : "T7",
+            "SD_CMD"  : "U6",
+            "SD_DAT3" : "Y5",
+            "SD_DAT2" : "AB5",
+            } # IOSTANDARD=LVCMOS33
+        ),
+
+    ]
+
+    def toolchain_prepare(self, fragment, name, **kwargs):
+        # Note: this is a setting that worked for the board used in bringup, does
+        # not mean it is a good setting for all process/voltage/temperature variations
+
+        # ToDo: input timing is not met according to timing report but it works,
+        # measure real timing with oscilloscope
+        extra_constraints = [
+
+            "NET \"ulpi_0__data__io<0>\" TNM = ulpi;",
+            "NET \"ulpi_0__data__io<1>\" TNM = ulpi;",
+            "NET \"ulpi_0__data__io<2>\" TNM = ulpi;",
+            "NET \"ulpi_0__data__io<3>\" TNM = ulpi;",
+            "NET \"ulpi_0__data__io<4>\" TNM = ulpi;",
+            "NET \"ulpi_0__data__io<5>\" TNM = ulpi;",
+            "NET \"ulpi_0__data__io<6>\" TNM = ulpi;",
+            "NET \"ulpi_0__data__io<7>\" TNM = ulpi;",
+            "NET \"forward_ulpi_clk_0__io\" TNM = ulpi;",
+            "NET \"ulpi_0__dir__io\"  TNM = ulpi;",
+            "NET \"ulpi_0__nxt__io\"  TNM = ulpi;",
+            "NET \"ulpi_0__stp__io\"  TNM = ulpi;",
+            "TIMEGRP \"ulpi\" OFFSET = OUT AFTER \"clk_100MHz_0__io\" REFERENCE_PIN \"forward_ulpi_clk_0__io\" RISING;",
+            "TIMEGRP \"ulpi\" OFFSET = IN 3 ns VALID 9 ns BEFORE \"clk_100MHz_0__io\" RISING;",
+
+        ]
+
+        overrides = { "add_constraints": "\n".join(extra_constraints) }
+        return super().toolchain_prepare(fragment, name, **overrides, **kwargs)
+
+    def toolchain_program(self, products, name):
```

Comment:
Is there any reason to implement an empty `toolchain_program` here? (And, out of curiosity, what tool is used to configure this board, normally?)

Reply:
I could not find a way to tell nmigen to stop after creation of build artifacts (emit_verilog()) so I tried to override some default handlers.

I think https://xkcd.com/763/ is relevant here. 

My workflow is nmigen/yosys -> verilog -> ISE GUI project -> manual programming via openocd (flyswater2 jtag connected to numato opsis via JTAG header P3).

My current workaround is this toplevel script:
`
import os
import sys

os.environ["LUNA_PLATFORM"] = "luna.gateware.platform.numato_opsis:NumatoOpsisPlatform"

os.environ["LUNA_ROOT"] = "<redacted>/luna_git"
#os.environ["LUNA_FULL_ONLY"] = "1"


sys.path.append(os.getenv('LUNA_ROOT') + "/examples/usb")
sys.path.append(os.getenv('LUNA_ROOT') + "/src/nmigen")
sys.path.append(os.getenv('LUNA_ROOT') + "/src/usb-protocol")
sys.path.append(os.getenv('LUNA_ROOT') + "/src/nmigen-boards")
sys.path.append(os.getenv('LUNA_ROOT') + "/src/nmigen-soc")
sys.path.append(os.getenv('LUNA_ROOT') + "/src/lambdasoc")

from luna.gateware.platform import get_appropriate_platform
from simple_device import USBDeviceExample

platform = get_appropriate_platform()

top = USBDeviceExample()

from nmigen      import Signal, ClockSignal
from nmigen.back import verilog
from nmigen.build import Subsignal, Pins

plan = platform.prepare(top)
plan.execute_local("build", run_script=False)
`

Reply:
Missed the "normally" part. The script provided by vanilla openocd installation suggest the normal way is:  download ixo-usb-jtag to the fx2 on the numato opsis and use "/usr/share/openocd/scripts/board/numato_opsis.cfg" for programming. But I have not tried it.